File: solution/[7453] 합이 0인 네 정수/손예진.cpp
#include <iostream>
#include <vector>
#include <algorithm>

//type overflow 주의하기

using namespace std;

vector<long long> first, second;
vector<int> A, B, C, D;
int N;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cin >> N;
	for (int i = 0; i < N; i++) {
		int a, b, c, d;
		cin >> a >> b >> c >> d;
		A.push_back(a);
		B.push_back(b);
		C.push_back(c);
		D.push_back(d);
	}
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			first.push_back(A[i] + B[j]);
			second.push_back(C[i] + D[j]);
		}
	}
	sort(first.begin(), first.end());
	sort(second.rbegin(), second.rend());
	int left = 0, right =  0;
	long long ans = 0;
	while (left < first.size() && right < second.size()) {
		long long sum = first[left] + second[right];
		if (sum == 0) {
			long long tmp1 = 1, tmp2 = 1;
			for (int i = left + 1; i < first.size(); i++, tmp1++) {
				if (first[left] != first[i])
					break;
			}
			for (int i = right + 1; i < second.size(); i++, tmp2++) {
				if (second[right] != second[i])
					break;
			}
			ans += tmp1 * tmp2;
			left += tmp1; right += tmp2;
		}
		else if (sum < 0)
			left++;
		else
			right++;
	}
	cout << ans << "\n";
}

File: solution/[7453] 합이 0인 네 정수/손예진.java
import java.util.*;

public class test_7453 {
    static int N, MAX;
    static int a[], b[], c[], d[], ab[], cd[];
    
    static long BinarySearch(){ //가지수가 최대 256000000000000개 나올 수 있음
        int left = 0, right = MAX - 1;
        long ans = 0;
        while(left < MAX && right>= 0){
            int sum = ab[left] + cd[right];
            if(sum == 0){    //합이 0인 경우
                long cnt1 = 1, cnt2 = 1;
                for(left++; left < MAX; left++){
                    if(ab[left] == ab[left - 1])
                        cnt1++;
                    else
                    	break;
                }
                for(right--; right >= 0; right--){
                    if(cd[right] == cd[right + 1])
                        cnt2++;
                    else
                    	break;
                }
                ans += cnt1  * cnt2;
            }
            else if(sum < 0){    //0보다 작으면 left를 키워야됨
                left++;
            }
            else{    //0보다 크면 right를 줄여야됨
                right--;
            }
        }
        return ans;
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        ab = new int[N * N];	//크기가 1 큰 상태로 sort 하면 0이 다른 값으로 하나 대체가 안됨
        cd = new int[N * N];
        a = new int[N];
        b = new int[N];
        c = new int[N];
        d = new int[N];
        MAX = N * N;
        for(int i = 0; i < N; i++){
            a[i] = sc.nextInt();
            b[i] = sc.nextInt();
            c[i] = sc.nextInt();
            d[i] = sc.nextInt();
        }
        int idx = 0;
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                ab[idx] = a[i] + b[j];
                cd[idx] = c[i] + d[j];
                idx++;
            }
        }
        Arrays.sort(ab);
        Arrays.sort(cd);
        System.out.println(BinarySearch());
    }
}
